File: internal/webui/assets/activity.js
// SPDX-License-Identifier: AGPL-3.0-or-later

// The activity log, read from the server one page at a time.
//
// This is the one view that does not work offline, and that is deliberate. The
// server holds the log, the client holds none of it, and a log a client cached
// would be stale the moment somebody else wrote a line. See
// internal/store/activity.go for why the table is outside sync.
//
// The store writes the fact and this file writes the sentence, exactly as
// internal/push does for a notification. See docs/DECISIONS.md D-020.

// VERBS turns an action into what a person reads. A row whose action is not
// here still draws, with the action itself as the words, so a command added to
// the server is never invisible here.
const VERBS = {
  task_add: 'added',
  task_update: 'changed',
  task_complete: 'completed',
  task_wont_do: 'gave up on',
  task_uncomplete: 'opened again',
  task_delete: 'deleted',
  task_restore: 'restored',
  task_move: 'moved',

  project_add: 'made the list',
  project_update: 'changed the list',
  project_delete: 'deleted the list',
  project_restore: 'restored the list',

  section_add: 'added the heading',
  section_update: 'renamed the heading',
  section_move: 'moved the heading',
  section_reorder: 'reordered the heading',
  section_delete: 'deleted the heading',
  section_restore: 'restored the heading',

  comment_add: 'commented on',
  comment_update: 'changed a comment on',
  comment_delete: 'deleted a comment on',

  label_add: 'added the label',
  label_delete: 'deleted the label',

  reminder_add: 'set a reminder on',
  reminder_update: 'changed a reminder on',
  reminder_delete: 'took a reminder off',

  login: 'signed in',
  login_failed: 'a sign-in failed',
  logout: 'signed out',
  passkey_add: 'added the passkey',
  passkey_delete: 'removed the passkey',
  invite_create: 'wrote an invitation for',
  invite_revoke: 'revoked an invitation',
  joined: 'joined the household',
  share: 'shared the list with',
  unshare: 'took the list back from',
};

// UNDO names the command that puts a deleted row back. A comment has none,
// because the server has no comment_restore, so the log does not offer one.
const UNDO = {
  task_delete: 'task_restore',
  project_delete: 'project_restore',
  section_delete: 'section_restore',
};

// SECURITY are the actions that belong to the audit half of the log. They read
// differently: nobody "did something to a task", and the address matters.
const SECURITY = new Set([
  'login', 'login_failed', 'logout', 'passkey_add', 'passkey_delete',
  'invite_create', 'invite_revoke', 'joined',
]);

export function verb(action) {
  return VERBS[action] || action.replace(/_/g, ' ');
}

export function isSecurity(action) {
  return SECURITY.has(action);
}

export function undoOf(action) {
  return UNDO[action] || '';
}

// fields turns "due_date,priority" into words a person reads.
const FIELDS = {
  due_date: 'the due date',
  due_time: 'the time',
  deadline: 'the deadline',
  start_date: 'the start date',
  priority: 'the priority',
  title: 'the title',
  description: 'the note',
  labels: 'the labels',
  assignee_id: 'who does it',
  section_id: 'the heading',
  project_id: 'the list',
  rrule: 'the repeat',
  duration_min: 'the duration',
};

export function detailWords(detail) {
  if (!detail) return '';
  const names = detail.split(',').filter(Boolean).map((f) => FIELDS[f] || f.replace(/_/g, ' '));
  if (!names.length) return '';
  if (names.length === 1) return names[0];
  return names.slice(0, -1).join(', ') + ' and ' + names[names.length - 1];
}

// ago says how long ago in the fewest words, the same way the comment panel
// does. A history is read as "this morning", never as a timestamp.
export function ago(stamp) {
  const then = Date.parse(stamp || '');
  if (!then) return '';
  const mins = Math.round((Date.now() - then) / 60000);
  if (mins < 1) return 'now';
  if (mins < 60) return mins + 'm ago';
  if (mins < 60 * 24) return Math.round(mins / 60) + 'h ago';
  const days = Math.round(mins / (60 * 24));
  if (days < 7) return days + 'd ago';
  return new Date(then).toLocaleDateString(undefined, { day: 'numeric', month: 'short' });
}

/**
 * page asks the server for one page of the log.
 *
 * scope is {project, task}, and before is the cursor: the smallest seq of the
 * page just read. It throws on a failure, and the caller says so on the screen:
 * an empty list would read as "nothing happened", which is a different fact.
 */
export async function page(scope, before, limit) {
  const p = new URLSearchParams();
  if (scope.project) p.set('project', scope.project);
  if (scope.task) p.set('task', scope.task);
  if (before) p.set('before', String(before));
  if (limit) p.set('limit', String(limit));
  const res = await fetch('/v1/activity?' + p.toString());
  if (!res.ok) throw new Error('the server answered ' + res.status);
  return res.json();
}

/**
 * draw paints the log into one element, and keeps painting as the reader asks
 * for more.
 *
 * ctx carries what this module must not own: esc for the markup, personName
 * for an account id, me for the account this browser is, restore to send the
 * undo command, gone to ask whether a row is still deleted, and open to show a
 * task.
 */
export function draw(box, scope, ctx) {
  let cursor = 0;
  let rows = [];

  const paint = (state) => {
    if (state === 'loading' && !rows.length) {
      box.innerHTML = '<span class="d-hint">Reading the history…</span>';
      return;
    }
    if (state === 'failed') {
      box.innerHTML = '<span class="d-hint">The history needs the network. '
        + 'Everything else here works offline.</span>';
      return;
    }
    if (!rows.length) {
      box.innerHTML = '<span class="d-hint">Nothing happened yet.</span>';
      return;
    }
    box.innerHTML = rows.map((r) => line(r, ctx)).join('')
      + (state === 'more' ? '<button class="acmore">Show more</button>' : '');

    const more = box.querySelector('.acmore');
    if (more) more.onclick = () => load();

    [...box.querySelectorAll('.acundo')].forEach((b) => {
      b.onclick = () => {
        ctx.restore(b.dataset.cmd, b.dataset.row);
        b.remove();
      };
    });
    [...box.querySelectorAll('.acopen')].forEach((b) => {
      b.onclick = () => ctx.open(b.dataset.task);
    });
  };

  const load = async () => {
    paint('loading');
    try {
      const d = await page(scope, cursor, 25);
      rows = rows.concat(d.activity || []);
      if (rows.length) cursor = rows[rows.length - 1].seq;
      paint(d.more ? 'more' : 'done');
    } catch (e) {
      paint(rows.length ? 'done' : 'failed');
    }
  };

  load();
}

// line draws one row. It never says "me did something": the first person reads
// as "you", which is how the comment panel already writes it.
function line(r, ctx) {
  const who = r.account_id === ctx.me ? 'You' : (r.actor || ctx.personName(r.account_id) || 'Someone');
  const what = ctx.esc(r.title || '');
  const undo = undoOf(r.action);
  const rowId = r.action.startsWith('task_') ? r.task_id : r.project_id;

  let tail = '';
  if (isSecurity(r.action)) {
    // A security line names the device and the address, and nothing else. It
    // has no task to open and nothing to restore.
    tail = [what, r.detail ? ctx.esc(r.detail) : '', r.addr ? ctx.esc(r.addr) : '']
      .filter(Boolean).map((x) => `<span class="acwhat">${x}</span>`).join(' ');
  } else if (r.task_id && ctx.known(r.task_id)) {
    tail = `<button class="acopen" data-task="${ctx.esc(r.task_id)}">${what}</button>`;
  } else {
    tail = `<span class="acwhat">${what}</span>`;
  }

  const fields = r.action === 'task_update' || r.action === 'project_update'
    ? detailWords(r.detail) : '';

  return `<div class="acrow">
    <span class="acwho">${ctx.esc(who)}</span>
    <span class="acverb">${ctx.esc(verb(r.action))}</span>
    ${tail}
    ${fields ? `<span class="acfields">— ${ctx.esc(fields)}</span>` : ''}
    <span class="acwhen">${ctx.esc(ago(r.at))}</span>
    ${undo && rowId && ctx.gone(r.action, rowId)
      ? `<button class="acundo" data-cmd="${undo}" data-row="${ctx.esc(rowId)}">Restore</button>` : ''}
  </div>`;
}

File: internal/store/activity.go
// SPDX-License-Identifier: AGPL-3.0-or-later

package store

import (
	"database/sql"
	"encoding/json"
	"sort"
	"strings"
)

// The activity log: who did what, and when.
//
// One table serves two readers. A person opens a project and reads its
// history, and the same rows are the audit trail that §6.6 of docs/PLAN.md
// asks for: every login, every token and every failed attempt.
//
// The store writes the fact and never the sentence, exactly as it does for a
// notification: a row holds an action name, an id and the title the row had at
// that moment, and the client writes the words. See docs/DECISIONS.md D-020.
//
// Recording happens in ONE place, right after a command succeeds, and not
// inside the fifty cases of applyOne. A log line that lives beside each case is
// a log line that the next case forgets.

// Activity is one line of the log.
type Activity struct {
	Seq       int64  `json:"seq"`
	At        string `json:"at"`
	AccountID string `json:"account_id"`
	// Actor is the name of the person, so a client never has to resolve an id
	// it may not be allowed to look up.
	Actor     string `json:"actor,omitempty"`
	Action    string `json:"action"`
	ProjectID string `json:"project_id,omitempty"`
	TaskID    string `json:"task_id,omitempty"`
	Title     string `json:"title,omitempty"`
	Detail    string `json:"detail,omitempty"`
	Addr      string `json:"addr,omitempty"`
}

// The actions that are not commands. A command writes its own type as the
// action, so only the security actions need names here.
const (
	ActionLogin         = "login"
	ActionLoginFailed   = "login_failed"
	ActionLogout        = "logout"
	ActionPasskeyAdd    = "passkey_add"
	ActionPasskeyDelete = "passkey_delete"
	ActionInviteCreate  = "invite_create"
	ActionInviteRevoke  = "invite_revoke"
	ActionJoined        = "joined"
	ActionShare         = "share"
	ActionUnshare       = "unshare"
)

// ActivityQuery asks for one page of the log.
//
// Before is the paging cursor: the answer holds rows below it, newest first,
// so the next page asks with the smallest seq it received. Zero means the
// newest page.
type ActivityQuery struct {
	ProjectID string
	TaskID    string
	Before    int64
	Limit     int
}

// maxActivityPage caps a page, so a client cannot ask for the whole log in one
// request.
const maxActivityPage = 200

// ActivityFor reads one page of the log that an account may see.
//
// A row is visible through its project, exactly as a task is. A row with no
// project is personal: a login, a token and a reminder belong to one person
// and nobody else reads them, not even the owner of the household.
func (s *Store) ActivityFor(q ActivityQuery, accountID string) ([]Activity, error) {
	limit := q.Limit
	if limit <= 0 || limit > maxActivityPage {
		limit = maxActivityPage
	}

	const mine = `(SELECT id FROM project WHERE owner_id = ?
		UNION SELECT project_id FROM project_member WHERE account_id = ?)`

	where := `(a.project_id IN ` + mine + ` OR (a.project_id IS NULL AND a.account_id = ?))`
	args := []any{accountID, accountID, accountID}

	if q.ProjectID != "" {
		where += ` AND a.project_id = ?`
		args = append(args, q.ProjectID)
	}
	if q.TaskID != "" {
		where += ` AND a.task_id = ?`
		args = append(args, q.TaskID)
	}
	if q.Before > 0 {
		where += ` AND a.seq < ?`
		args = append(args, q.Before)
	}
	args = append(args, limit)

	rows, err := s.db.Query(`SELECT a.seq, a.at, a.account_id,
		coalesce(nullif(c.display_name, ''), c.name, ''),
		a.action, coalesce(a.project_id, ''), coalesce(a.task_id, ''), a.title, a.detail, a.addr
		FROM activity a LEFT JOIN account c ON c.id = a.account_id
		WHERE `+where+` ORDER BY a.seq DESC LIMIT ?`, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	out := []Activity{}
	for rows.Next() {
		var a Activity
		if err := rows.Scan(&a.Seq, &a.At, &a.AccountID, &a.Actor, &a.Action,
			&a.ProjectID, &a.TaskID, &a.Title, &a.Detail, &a.Addr); err != nil {
			return nil, err
		}
		out = append(out, a)
	}
	return out, rows.Err()
}

// Note records one action that is not a command: a login, a passkey, an
// invitation. It is the audit half of §6.6.
//
// It fails quietly on purpose. A login that works must not fail because the
// log could not be written, and a login that is refused must still be refused.
// The caller has nothing useful to do with the error either way.
func (s *Store) Note(accountID, action, title, detail, addr string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	_, _ = s.db.Exec(`INSERT INTO activity (at, account_id, action, title, detail, addr)
		VALUES (?,?,?,?,?,?)`, s.stamp(), accountID, action, title, detail, addr)
}

// NoteIn records an action against one project, so the other people who share
// the list read it too. Sharing a list is the example: it is not a command,
// and it is not private either.
func (s *Store) NoteIn(accountID, action, projectID, title, detail string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	_, _ = s.db.Exec(`INSERT INTO activity (at, account_id, action, project_id, title, detail)
		VALUES (?,?,?,?,?,?)`, s.stamp(), accountID, action, projectID, title, detail)
}

// recordActivity writes the log line for a command that has just succeeded.
//
// It runs inside the command's transaction, after the savepoint was released,
// so a command that failed leaves no line. It reads the row back rather than
// trusting the arguments: a task_add that took its project from a name, and a
// task_update that changed nothing, both have to log where the row actually
// is.
//
// It never fails the command. A log that cannot be written is a lost line, and
// losing the write itself is worse.
func recordActivity(tx *sql.Tx, c Command, id, now string, act actor) {
	if id == "" {
		return
	}
	var projectID, taskID, title sql.NullString
	detail := ""

	switch {
	case strings.HasPrefix(c.Type, "task_"):
		taskID = sql.NullString{String: id, Valid: true}
		var p, t string
		if err := tx.QueryRow(`SELECT project_id, title FROM task WHERE id = ?`, id).Scan(&p, &t); err != nil {
			return
		}
		projectID = sql.NullString{String: p, Valid: true}
		title = sql.NullString{String: t, Valid: true}
		if c.Type == "task_update" {
			detail = changedFields(c.Args)
		}

	case strings.HasPrefix(c.Type, "project_"):
		var name string
		if err := tx.QueryRow(`SELECT name FROM project WHERE id = ?`, id).Scan(&name); err != nil {
			return
		}
		projectID = sql.NullString{String: id, Valid: true}
		title = sql.NullString{String: name, Valid: true}
		if c.Type == "project_update" {
			detail = changedFields(c.Args)
		}

	case strings.HasPrefix(c.Type, "section_"):
		var p, name string
		if err := tx.QueryRow(`SELECT project_id, name FROM section WHERE id = ?`, id).Scan(&p, &name); err != nil {
			return
		}
		projectID = sql.NullString{String: p, Valid: true}
		title = sql.NullString{String: name, Valid: true}

	case strings.HasPrefix(c.Type, "comment_"):
		// The line is logged, and its text is not. A comment is a row that a
		// client can read, and a deleted comment whose words survive in a log
		// nobody can delete is not what a person means by deleting it.
		var t string
		if err := tx.QueryRow(`SELECT task_id FROM comment WHERE id = ?`, id).Scan(&t); err != nil {
			return
		}
		var p, name string
		if err := tx.QueryRow(`SELECT project_id, title FROM task WHERE id = ?`, t).Scan(&p, &name); err != nil {
			return
		}
		taskID = sql.NullString{String: t, Valid: true}
		projectID = sql.NullString{String: p, Valid: true}
		title = sql.NullString{String: name, Valid: true}

	case strings.HasPrefix(c.Type, "reminder_"):
		// A reminder belongs to one person even on a shared task, so its line
		// carries no project and only its owner reads it.
		var t sql.NullString
		if err := tx.QueryRow(`SELECT task_id FROM reminder WHERE id = ?`, id).Scan(&t); err != nil {
			return
		}
		taskID = t
		if t.Valid && t.String != "" {
			var name string
			if err := tx.QueryRow(`SELECT title FROM task WHERE id = ?`, t.String).Scan(&name); err == nil {
				title = sql.NullString{String: name, Valid: true}
			}
		}

	case strings.HasPrefix(c.Type, "label_"):
		// A label is household vocabulary and belongs to no project, so its
		// line is personal to whoever wrote it.
		var name string
		if err := tx.QueryRow(`SELECT name FROM label WHERE id = ?`, id).Scan(&name); err != nil {
			return
		}
		title = sql.NullString{String: name, Valid: true}

	default:
		return
	}

	_, _ = tx.Exec(`INSERT INTO activity (at, account_id, action, project_id, task_id, title, detail)
		VALUES (?,?,?,?,?,?,?)`, now, act.ID, c.Type, projectID, taskID, title.String, detail)
}

// changedFields names the fields an update command asked to change, in one
// sorted comma-separated string.
//
// It reads the argument object rather than a typed struct, so a field added to
// a command appears in the log with no change here. `clear` is expanded,
// because a person reads "the deadline went away" and not "a clear happened".
func changedFields(raw json.RawMessage) string {
	var m map[string]json.RawMessage
	if err := json.Unmarshal(raw, &m); err != nil {
		return ""
	}
	seen := map[string]bool{}
	if c, ok := m["clear"]; ok {
		var names []string
		if err := json.Unmarshal(c, &names); err == nil {
			for _, n := range names {
				seen[n] = true
			}
		}
	}
	for k := range m {
		if k == "id" || k == "clear" {
			continue
		}
		seen[k] = true
	}
	keys := make([]string, 0, len(seen))
	for k := range seen {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return strings.Join(keys, ",")
}
